keyserver/listener.go

```diff
@@ -0,0 +1,140 @@
+package keyserver
+
+import (
+	"encoding/json"
+	"io"
+	"log"
+	"net"
+	"time"
+
+	. "github.com/coniks-sys/coniks-go/protocol"
+)
+
+func (server *ConiksServer) listenForRequests(ln net.Listener, handler func(msg []byte) ([]byte, error)) {
+	defer ln.Close()
+	go func() {
+		<-server.stop
+		ln.Close()
+	}()
+
+	for {
+		conn, err := ln.Accept()
+		if err != nil {
+			select {
+			case <-server.stop:
+				return
+			default:
+				log.Printf("accept client: %s", err)
+				continue
+			}
+		}
+		server.waitStop.Add(1)
+		go func() {
+			server.acceptClient(conn, handler)
+			server.waitStop.Done()
+		}()
+	}
+}
+
+func (server *ConiksServer) acceptClient(conn net.Conn, handler func(msg []byte) ([]byte, error)) {
+	defer conn.Close()
+	closed := make(chan struct{})
+	defer close(closed)
+	go func() {
+		select {
+		case <-server.stop:
+			conn.Close()
+		case <-closed:
+		}
+	}()
+
+	conn.SetDeadline(time.Now().Add(5 * time.Second))
+
+	// handle request
+	buf := make([]byte, 4096)
+	for {
+		n, err := conn.Read(buf)
+		if err != nil {
+			select {
+			case <-server.stop:
+				return
+			default:
+				if err != io.EOF {
+					log.Printf("client read %v: %v", conn.RemoteAddr(), err)
+				}
+				return
+			}
+		}
+
+		res, err := handler(buf[:n])
+		if err != nil {
+			log.Printf("client handle %v: %v", conn.RemoteAddr(), err)
+		}
+
+		n, err = conn.Write([]byte(res))
+		if err != nil {
+			select {
+			case <-server.stop:
+				return
+			default:
+				log.Printf("client write %v: %v", conn.RemoteAddr(), err)
+				return
+			}
+		}
+	}
+}
+
+func malformedClientMsg() []byte {
+	res, err := MarshalErrorResponse(ErrorMalformedClientMessage)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+// handleClientMessage returns a byte slice of marshaled response
+// and an error for server logging
+func (server *ConiksServer) handleClientMessage(msg []byte) ([]byte, error) {
+	// get request message
+	req, _, err := UnmarshalRequest(msg)
+	if err != nil {
+		return malformedClientMsg(), err
+	}
+
+	// handle request
+	switch req.Type {
+	default:
+		log.Printf("unknown message type: %q", req.Type)
+		return malformedClientMsg(), ErrorMalformedClientMessage.Error()
+	}
+}
+
+func (server *ConiksServer) handleBotMessage(msg []byte) ([]byte, error) {
```

Comment:
What do you think of this ... combine `handleBotMessage` and `handleClientMessage` into one func, but you make another function that binds the `req.Type`s it accepts.

```
func makeHandler(acceptableTypes []int)  func(msg []byte) ([]byte, error) {
  return func(msg []byte) { ... }
}
```